[docs] Update migration guide to have a section on nested classes (#27354)
mui-org/material-ui: docs/src/pages/guides/migration-v4/migration-v4.md

  +import { DistributiveOmit } from '@material-ui/types';
  ```

## Migrate theme's `styleOverrides` to emotion

> Although your style overrides defined in the theme may partially work, there is an important difference on how the nested elements are styled. The `$` syntax used with JSS will not work with emotion. You need to replace those selectors with a valid class selectors.

Reviewer: ```suggestion
Although your style overrides defined in the theme may partially work, there is an important difference on how the nested elements are styled. The `$` syntax used with JSS will not work with Emotion. You need to replace those selectors with a valid class selector.
```

Typos and branding nits

### Replace pseudo state class names

```diff
const theme = createTheme({
  components: {
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
-         '&$focused': {
+         '&.Mui-focused': {
            borderWidth: 1,
          }
        }
      }
    }
  }
});
```

### Replace nested classes selectors with global class names

```diff
const theme = createTheme({
  components: {
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
-         '& $notchedOutline': {
+         '& .MuiOutlinedInput-notchedOutline': {
            borderWidth: 1,
          }
        }
      }
    }
  }
});
```

> > Note: For each component we export `[component]Classes` constant that contains all nested classes for that component. You can rely on this instead of hardcoding the classes.

Reviewer: ```suggestion
> Note: For each component we export a `[component]Classes` constant that contains all nested classes for that component. You can rely on this instead of hardcoding the classes.
```

```diff
+import { outlinedInputClasses } from '@material-ui/core/OutlinedInput';

const theme = createTheme({
  components: {
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
-         '& $notchedOutline': {
+         [`& .${outlinedInputClasses['notchedOutline']}`]: {
            borderWidth: 1,
          }
        }
      }
    }
  }
});
```

> Take a look on the whole [list of the pseudo-state global classname](/customization/how-to-customize/#pseudo-classes) available.

Reviewer: ```suggestion
Take a look at the whole [list of pseudo-state global classnames](/customization/how-to-customize/#pseudo-classes) available.
```
